department-of-veterans-affairs/vets-website
# Use IDs to refer to disabilities

## Description
I could break this up into two PRs (one for adding the `updateDataHooks` and one for using them), but I didn't think one would be useful without the other, so I opted to keep it all together. Adding the hooks into the forms library was pretty low-touch.

I think I regret that decision. Sorry. :grimacing: 

This PR does a few things:
1) Adds the ability to put in page-config-level `formData` update hooks
2) Calls them in the form flow and review page
3) The `newDisabilities` page adds a hook that handles adding and removing new disabilities
   - New disabilities get a uuid
   - Removed disabilities also get deleted from `treatedDisabilityNames` and `powDisabilities`
   - Changed disabilities don't need any updating in the form data since we now use IDs for relationships
4) Adds a new migration that
   - Gives all rated and new disabilities a uuid
   - Uses the uuid in `treatedDisabilityNames` and `powDisabilities`
5) Gives rated disabilities a uuid in the pre-fill transformer
   - This coupled with the `updateDataHook` means all disabilities will have a uuid
6) The submit transformer
   - Translates the `treatedDisabilityNames` and `powDisabilities` into an array of the proper disability names using the uuid as a lookup key
   - Removes the uuid from the disabilities
     - I originally had it as a `view:uuid`, but I needed it after the `view:` fields were stripped out in the submit transformer


## Testing done
Testing strategy:
- Before checking out the branch, fill out a form with:
  - Multiple new disabilities
  - VA medical records
  - POW status
- Get to the review page and save the form
- Check out this branch
- Load the saved form and ensure the right disabilities are checked
  off in both the VA treatment facilities and POW disabilities
- Remove one of the disabilities
- Change the name of another disability
- Ensure the one that was removed is gone from `treatedDisabilityNames` and `powDisabilities`
- Ensure that the changed one's name is updated in both lists

This tests the review page. You can go back to the page in the form
flow and do the same thing (removing and renaming) to test the code in
`FormPage`.

I tested this manually and wrote a whole lotta tests.

## TODO
- Update the forms library documentation (if what we have currently is up-to-date)
  - I'd kind of prefer to do this in another PR because this one is already pretty massive.

## Screenshots
N/A

## Acceptance criteria
- [x] The migration is run to add ids to saved forms
- [x] The `updateDataHooks` are run to add ids to new disabilities and remove entries from `treatedDisabilityNames` and `powDisabilities` when appropriate
- [x] Pre-fill adds ids to rated disabilities
- [x] The output of the submit transformer remains the same

## Definition of done
